text/0000-erfc-post-build-contexts.md

```diff
@@ -0,0 +1,396 @@
+these tools usually operate at a different layer of abstraction so it might not be necessary.
+
+A major drawback of this proposal is that it is very general, and perhaps too powerful. We're currently using the
+more focused API in the eRFC, and may switch to this during experimentation if a pressing need crops up.
+
+## Alternative procedural macro with minimal compiler changes
+
+The above proposal can be made even more general, minimizing the impact on the compiler.
+
+This assumes that `#![foo]` ("inner attribute") macros work on modules and on crates.
+
+The idea is that the compiler defines no new proc macro surface, and instead simply exposes
+a `--attribute` flag. This flag, like `-Zextra-plugins`, lets you attach a proc macro attribute
+to the whole crate before compiling. (This flag actually generalizes a bunch of flags that the
+compiler already has)
+
+Test crates are now simply proc macro attributes:
+
+```rust
+#[proc_macro_attr(attributes(test, foo, bar))]
+pub fn harness(crate: TokenStream) -> TokenStream {
+  // ...
+}
+```
+
+The cargo functionality will basically compile the file with the right dependencies
+and `--attribute=your_crate::harness`.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+These are mostly intended to be resolved during the experimental
+feature. Many of these have strawman proposals -- unlike the rest of this RFC,
+these proposals have not been discussed as thoroughly. If folks feel like
+there's consensus on some of these we can move them into the main RFC.
+
+## Integration with doctests
+
+Documentation tests are somewhat special, in that they cannot easily be
+expressed as `TokenStream` manipulations. In the first instance, the
+right thing to do is probably to have an implicitly defined post-build
+context called `doctest` which is included in the post-build context set
+`test` by default (as proposed above).
+
+Another argument for punting on doctests is that they are intended to
+demonstrate code that the user of a library would write. They're there
+to document *how* something should be used, and it then makes somewhat
+less sense to have different "ways" of running them.
+
+## Translating existing cargo test flags
+
+Today, `cargo test` takes a number of flags such as `--lib`, `--test
+foo`, and `--doc`. As it would be a breaking change to change these,
+cargo should recognize them and map to the appropriate post-build
+contexts.
+
+Furthermore, `cargo test` lets you pick a single testing target via `--test`,
+and `cargo bench` via `--bench`. We'll need to create an agnostic flag
+for `cargo context` (we cannot use `--target` because it is already used for
+the target architecture, and `--test` is too specific for tests). `--post-build-target`
+is one rather verbose suggestion.
+
+We also need to settle on a command name, `cargo context` and `cargo post-build`
+don't quite capture what's going on.
+
```

Comment:
FWIW --test is internally copied over as the first argument to the test binary, so this can be handled in `main()`. But yeah, this could be useful.

I'd leave this up to the experimental phase to work in IMO.